### PR TITLE
Flag invalid Navigation Link items.

### PR DESCRIPTION
**Pull request by vcanales**
## Description

Re. https://github.com/WordPress/gutenberg/issues/23573

Following the criteria used when deciding whether or not to render a Navigation Link item server-side, mark those items that would not be rendered as Invalid in the editor, in order to match the behaviour of the menu editor found on `nav-menus.php`.

The steps taken to decide whether or not an element is invalid are the following:
- Check Navigation Link validity if:
1. Link is 'post-type'.
2. It has an id.
3. It's neither null, nor undefined, as valid items might be either of those while loading.

If those conditions are met, check if
1. The post status is published.
2. The Navigation Link item has no label.

If either of those is true, invalidate.

## How has this been tested?
Manually tested by:
1. Creating a few pages.
2. Creating a Navigation Menu with all pages.
3. Deleting one of those pages.
4. Setting another one of those pages to 'Draft' status.

Both modified pages should now be flagged with the text "(Invalid)".

And
1. Creating a few posts.
2. Adding a Post Links pointing to those post.
3. Setting some of those posts to Draft.
4. Trashing some of those posts.

All Drafts and trashed posts should be flagged with the text "(Invalid)".

## Screenshots 

image

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by jasmussen** [comment]
This is cool:

Screenshot 2021-05-12 at 12 10 25

I pushed a small padding change:

Screenshot 2021-05-12 at 12 11 12

I think we can potentially revisit and refine the menu item placeholder style as we go, but that's something I'll be looking into separately. This just needs a code review from what I can tell! Nice.

**Comment by vcanales** [comment]
Thanks for taking a look. These changes broke the e2e tests, which I think indicates that some of those tests depend on invalid links showing up normally.

> I think we can potentially revisit and refine the menu item placeholder style as we go, but that's something I'll be looking into separately. This just needs a code review from what I can tell! Nice.

Agreed, I'd appreciate a look from @WordPress/gutenberg-core

File: packages/block-library/src/navigation-link/edit.js [hunk 1]
```diff
@@ -242,6 +243,37 @@ export default function NavigationLinkEdit( {
 	const itemLabelPlaceholder = __( 'Add link…' );
 	const ref = useRef();
 
+	const isPostType =
+		( kind && 'post-type' === kind ) || ( type && 'post' === type );
+
+	const hasId = id && ! isNaN( id );
```
Comment (ntsekouras) [inline packages/block-library/src/navigation-link/edit.js]: Why the `isNaN` check is needed?

File: packages/block-library/src/navigation-link/edit.js [hunk 2]
```diff
@@ -242,6 +243,37 @@ export default function NavigationLinkEdit( {
...
+	const postStatus = useSelect( ( select ) => {
+		if ( isPostType ) {
+			const { getEntityRecord } = select( coreStore );
+			const entityRecord = getEntityRecord( 'postType', type, id );
+
+			return entityRecord?.status;
+		}
+
+		return null;
+	} );
```
Comment (ntsekouras) [inline packages/block-library/src/navigation-link/edit.js]: We need to add the dependencies here: `[ isPostType, type, id ]`. Also let's revert the `if` here.

File: packages/block-library/src/navigation-link/edit.js [hunk 3]
```diff
@@ -242,6 +243,37 @@ export default function NavigationLinkEdit( {
...
+
+	// Check Navigation Link validity if:
+	// 1. Link is 'post-type'.
+	// 2. It has an id.
+	// 3. It's neither null, nor undefined, as valid items might be either of those while loading.
+	// If those conditions are met, check if
+	// 1. The post status is published.
+	// 2. The Navigation Link item has no label.
+	// If either of those is true, invalidate.
+	const isInvalid =
```
Comment (ntsekouras) [inline packages/block-library/src/navigation-link/edit.js]: This feels like it could be a bit more pleasant to the eyes 😄 
Maybe something like that:
```
const isInvalid =
	isPostType &&
	hasId &&
	postStatus &&
	( postStatus !== 'publish' || ! label );
```
Didn't give it much thought, so it could become better.

**Review by ntsekouras — COMMENTED** [review]
Thanks! I tested and works as expected. I didn't test with CPTs though..

Can we extract this in a separate hook? What do you think?

File: packages/block-library/src/navigation-link/edit.js (hunk 1 above)
Comment (vcanales) [reply packages/block-library/src/navigation-link/edit.js]: I am not sure of that myself — I added it in an attempt to keep this check consistent with what's done in index.php to decide whether or not a link should be rendered:

https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation-link/index.php#L106-L109

File: packages/block-library/src/navigation-link/edit.js (hunk 3 above)
Comment (vcanales) [reply packages/block-library/src/navigation-link/edit.js]: I don't think it gets much better than that.
I removed the label check though, as it doesn't make sense in the editor — it would, for instance, make it impossible to delete the label on a page link in order to edit it.

File: packages/block-library/src/navigation-link/edit.js (hunk 1 above)
Comment (ntsekouras) [reply packages/block-library/src/navigation-link/edit.js]: I changed this to `const hasId = Number.isInteger( id );`. It should be enough for what we are checking here.

**Review by ntsekouras — COMMENTED** [review]
I rebased but the Navigation failing tests might be relevant or related to `core` changes 🤔 . Needs investigation

One thing that we might take into account is that while the `link` is updated to `invalid link`, in ListView is still shown as before. Should we handle this views as well?

Screenshot 2021-05-25 at 10 54 40 AM

File: packages/block-library/src/navigation-link/edit.js [hunk 4]
```diff
@@ -520,6 +550,12 @@ export default function NavigationLinkEdit( {
 								} }
 							/>
 							{ missingText }
+							{ isInvalid && (
+								<span className="wp-block-navigation-link__invalid-item">
+									{ ' ' }
+									({ __( 'Invalid' ) })
```
Comment (gwwar) [inline packages/block-library/src/navigation-link/edit.js]: Since this is a pretty short string, I think leaving a translator note would be useful

File: packages/block-library/src/navigation-link/edit.js [hunk 5]
```diff
@@ -209,6 +209,34 @@ export const updateNavigationLinkBlockAttributes = (
 	} );
 };
 
+const useIsInvalidLink = ( kind, type, id ) => {
+	const isPostType = kind === 'post-type' || type === 'post';
```
Comment (gwwar) [inline packages/block-library/src/navigation-link/edit.js]: I think we'll also want to check for type of 'page'. This issue would only show up for folks with old data since kind is not set.

```jsx
const isPostType = kind === 'post-type' || type === 'post' || type = 'page';
```

It looks like we should be mirroring checks in https://github.com/WordPress/gutenberg/blob/efc28f91fb2c4924b173bc7c8f60dbffac62a7d2/packages/block-library/src/navigation-link/index.php#L107-L112

File: packages/block-library/src/navigation-link/edit.js [hunk 6]
```diff
@@ -520,6 +550,12 @@ export default function NavigationLinkEdit( {
 								} }
 							/>
 							{ missingText }
+							{ isInvalid && (
+								<span className="wp-block-navigation-link__invalid-item">
+									{ ' ' }
```
Comment (gwwar) [inline packages/block-library/src/navigation-link/edit.js]: What does the extra space here do?

**Review by gwwar — COMMENTED** [review]
Overall, this looks really solid! Nice work @vcanales ✨  As discussed in chat it looks like e2e mocks need to be updated to handle the new `getEntityRecord` call to fix tests.

One usability question I had was around how do folks know why this is invalid and how to fix it? Current workflows allow us to create drafts within the search suggestion, so seeing this from the start may be surprising.

https://user-images.githubusercontent.com/1270189/121251164-6ced4400-c85b-11eb-9600-96da1f28f8db.mp4

File: packages/block-library/src/navigation-link/edit.js (hunk 6 above)
Comment (vcanales) [reply packages/block-library/src/navigation-link/edit.js]: When concatenating text, span elements, or a combination of both, it's hard to keep spaces consistent, even more so when formatting breaks a `span` into mutiple lines.

The current way this is being handled is hard to read, so I've put the `missing text` logic in a hook that also concatenates `invalid text` when appropriate.

File: packages/block-library/src/navigation-link/edit.js (hunk 5 above)
Comment (vcanales) [reply packages/block-library/src/navigation-link/edit.js]: Those are exactly the checks I meant to mirror, must have missed that one. Thanks!

File: .vscode/launch.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (tellthemachines) [inline .vscode/launch.json]: Should this file be committed?

File: packages/block-library/src/navigation-link/edit.js [hunk 8]
```diff
@@ -210,6 +210,79 @@ export const updateNavigationLinkBlockAttributes = (
...
+		<>
+			<span>{ missingText }</span>
+			{ isInvalid && (
+				<span className="wp-block-navigation-link__invalid-item">
+					{ ' ' }
+					(
+					{
+						__(
+							'Invalid'
+						) /* translators: Whether or not the navigation link is invalid. */
```
Comment (tellthemachines) [inline packages/block-library/src/navigation-link/edit.js]: I'm not 100% sure the translators comment will be picked up in this position. Usually it's on the line immediately above the string, would be good to confirm if having it after works too.

**Review by tellthemachines — COMMENTED** [review]
Thanks for working on this! A couple minor questions below, and some thoughts about the Draft status:

The current menus screen only flags deleted posts/pages as invalid. Drafts are kept in the menu:

Screen Shot 2021-07-08 at 1 52 53 pm

The front end behaviour also differs with current menu showing draft pages (though they link to a 404 unless user is logged in/able to access the draft), while the Navigation block hides them. 

I think unless we have really good reasons to change this, we should follow current behaviour so users know what to expect. This would also help solve @gwwar 's point about creating draft pages in the Navigation block.

**Comment by vcanales** [comment]
Apologies for the delay here, it took me a long time to stabilize e2e tests...

> I think unless we have really good reasons to change this, we should follow current behaviour so users know what to expect. This would also help solve @gwwar 's point about creating draft pages in the Navigation block.

I've considered this, and made the change so that only trashed pages will be considered invalid.
I also think it still makes a lot of sense to mark draft pages, so I've added that.

Kapture 2021-08-25 at 14 34 00

If this makes sense, I'd love some input from @jasmussen to make it look presentable.

> One usability question I had was around how do folks know why this is invalid and how to fix it? Current workflows allow us to create drafts within the search suggestion, so seeing this from the start may be surprising.

With Draft pages being handled, I think it would make sense to look into hinting the user as to why the link is Invalid and how to fix it, as @gwwar suggests; this would improve both the Nav Block and the Nav Menu Editor experience; perhaps a follow up PR would be reasonable in this case.

**Comment by jasmussen** [comment]
Awesome work, this is what I see:

this

This is close, honestly. Two things come to mind. First off, for top level links, the "(Draft)" word has a little space before it — cool — but that space is missing in the submenu item:

Screenshot 2021-09-01 at 18 56 26

That space should probably just be actual literal whitespace in the code (` `), and it'd work fine. Can even sit inside the span and it'd be good.

The other issue is what happens if you link to a page, then trash that page and reload. In that case, the menu item reverts to a partial setup state, one that _has the link, but not the page title anymore_. I'm not sure if it's possible to keep the label intact when the target item is trashed? Right now the label reverts to "Select page", which paired with an existing link isn't very useful.

_Tangentially, #33091 suggests the reverse be possible, a label but no link._

Note, there are some appender alignment issues, which I'm pretty sure get solved with a rebase. Probably worth it just to be sure.

**Comment by jasmussen** [comment]
Here's a mockup of how I'd love it to look for the invalid item:

Screenshot 2021-09-03 at 11 21 39

Essentially what you have now, except the label is kept. Possible?

**Comment by vcanales** [comment]
> Essentially what you have now, except the label is kept. Possible?

Yes! Tried in on the latest commit https://github.com/WordPress/gutenberg/pull/31716/commits/b517dba6e0932eb91ed38d9298ad172bc9f477b9. The only issue I see right now is that `(Invalid)` gets wrapped. I can look into how to keep it in one line, but how fast should we allow it to extend? 

image

**Comment by jasmussen** [comment]
Nice, that feels very close. So, the page label of the item that's been trashed and is invalid, it doesn't have to be editable until the item becomes valid again. Looking at this chunk of code:

```
							<RichText
								ref={ ref }
								identifier="label"
								className="wp-block-navigation-link__label"
								value={ label }
								onChange={ ( labelValue ) =>
									setAttributes( { label: labelValue } )
								}
								onMerge={ mergeBlocks }
								onReplace={ onReplace }
								__unstableOnSplitAtEnd={ () =>
									insertBlocksAfter(
										createBlock( 'core/navigation-link' )
									)
								}
								aria-label={ __( 'Navigation link text' ) }
								placeholder={ itemLabelPlaceholder }
								withoutInteractiveFormatting
								allowedFormats={ [
									'core/bold',
									'core/italic',
									'core/image',
									'core/strikethrough',
								] }
								onClick={ () => {
									if ( ! url ) {
										setIsLinkOpen( true );
									}
								} }
							/>
							{ isInvalid && (
								<div className="wp-block-navigation-link__placeholder-text">
									<KeyboardShortcuts
										shortcuts={ {
											enter: () =>
												isSelected &&
												setIsLinkOpen( true ),
										} }
									/>
									<span>
										{ ' ' }
										(
										{ __(
											/* translators: Whether or not the navigation link is invalid. */
											'Invalid'
										) }
										)
									</span>
								</div>
							) }
```

that outputs two divs. One for the rich text editable, and `wp-block-navigation-link__placeholder-text`. 

In a quick test, I changed that to this:

```
							{ isInvalid && (
								<div className="wp-block-navigation-link__placeholder-text">
									<KeyboardShortcuts
										shortcuts={ {
											enter: () =>
												isSelected &&
												setIsLinkOpen( true ),
										} }
									/>
									<span>
										{ label } (
										{ __(
											/* translators: Whether or not the navigation link is invalid. */
											'Invalid'
										) }
										)
									</span>
								</div>
							) }
```

that seems to solve it. What do you think?

**Comment by jasmussen** [comment]
The code sample above I think is a step on the path as it removes the editable rich text from the invalid link, making it inert until it has been linked to a page again. However it appears to also remove the actual rich text from valid links. So we probably need a `{!invalid ... }` wrapper around the RichText part. What do you think @vcanales?

**Comment by vcanales** [comment]
Thanks for the review! After a bit of tinkering with it, it's now only displaying the Rich Text component if the link is not marked as a draft, or invalid.

Kapture 2021-09-06 at 14 31 48

This change allowed me to remove some repetition from the invalid link render code.

**Comment by jasmussen** [comment]
Nice work. I pushed a tiny fix in https://github.com/WordPress/gutenberg/pull/31716/commits/1af55a6d81e97e7bd67f3cc03ca7460beb7bcb4c that adds an extra CSS class to the placeholder item, that helps the layout stay intact. Before, if you had a submenu on an invalid item, you'd get this:

Screenshot 2021-09-06 at 19 55 32

With the fix pushed, that's fixed:

Screenshot 2021-09-06 at 19 59 49

**Review by jasmussen — APPROVED** [review]
Giving tentative approval as this is working solidly for me. I would encourage another sanity check if you can get it. Nice work!

**Comment by spacedmonkey** [comment]
Worth noting that if you are using the menu item request rest api endpoint, there is a field called `invalid`. This handles all the checking for post / page / term exists for you on the server.

File: packages/block-library/src/navigation-link/edit.js [hunk 9]
```diff
@@ -265,6 +266,64 @@ export const updateNavigationLinkBlockAttributes = (
...
+		case 'tag':
+			/* translators: label for missing tag in navigation link block */
+			missingText = __( 'Select tag' );
+			break;
+		default:
+			/* translators: label for missing values in navigation link block */
+			missingText = __( 'Add link' );
+	}
+
+	return <span>{ missingText }</span>;
```
Comment (spacedmonkey) [inline packages/block-library/src/navigation-link/edit.js]: I don't love that these are hard coded. Can we use taxonomy / post types for these translations?

File: packages/block-library/src/navigation-link/edit.js [hunk 10]
```diff
@@ -265,6 +266,64 @@ export const updateNavigationLinkBlockAttributes = (
...
+	// If those conditions are met, check if
+	// 1. The post status is published.
+	// 2. The Navigation Link item has no label.
+	// If either of those is true, invalidate.
+	const isInvalid =
+		isPostType && hasId && postStatus && 'trash' === postStatus;
+	const isDraft = 'draft' === postStatus;
+
+	return [ isInvalid, isDraft ];
+};
```
Comment (spacedmonkey) [inline packages/block-library/src/navigation-link/edit.js:297]: This should not be needed. There is a field in the REST API for this. 

https://github.com/WordPress/gutenberg/blob/9c3c53bc75b02be69d25f9cf3bae400f7dba7c10/lib/class-wp-rest-menu-items-controller.php#L1004-L1009

File: packages/block-library/src/navigation-link/edit.js [hunk 11]
```diff
@@ -265,6 +266,64 @@ export const updateNavigationLinkBlockAttributes = (
 	} );
 };
 
+const useIsInvalidLink = ( kind, type, id ) => {
+	const isPostType =
+		kind === 'post-type' || type === 'post' || type === 'page';
```
Comment (spacedmonkey) [inline packages/block-library/src/navigation-link/edit.js:271]: Also, custom post types and taxonomies?

File: packages/block-library/src/navigation-link/edit.js (hunk 10 above)
Comment (vcanales) [reply packages/block-library/src/navigation-link/edit.js:297]: It doesn't seem to be included in the return value of `getEntityRecord`; how can I access that property?

File: packages/block-library/src/navigation-link/edit.js (hunk 9 above)
Comment (vcanales) [reply packages/block-library/src/navigation-link/edit.js]: This logic was only moved in the context of this PR, not added, so I'd address this concern on a different PR.

File: packages/block-library/src/navigation-link/edit.js (hunk 11 above)
Comment (tellthemachines) [reply packages/block-library/src/navigation-link/edit.js:271]: We don't flag deleted taxonomies as invalid in classic menus, we just auto-remove them. We could look at doing the same for nav blocks, but perhaps as a separate task.

File: packages/block-library/src/navigation-link/edit.js (hunk 9 above)
Comment (tellthemachines) [inline packages/block-library/src/navigation-link/edit.js]: Do we need to wrap missingText in a span here? It's already output in a span below, so this is causing it to be double-wrapped.

File: packages/block-library/src/navigation-link/edit.js [hunk 12]
```diff
@@ -685,38 +730,68 @@ export default function NavigationLinkEdit( {
...
+										shortcuts={ {
+											enter: () =>
+												isSelected &&
+												setIsLinkOpen( true ),
+										} }
+									/>
+									<Tooltip
+										position="top center"
+										text={ __(
+											'This item has been deleted, or is a draft'
```
Comment (tellthemachines) [inline packages/block-library/src/navigation-link/edit.js]: I'm wondering how we could make this message screen-reader accessible. Perhaps we could add it in a hidden span and reference it from the div with `aria-describedby`?

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js [hunk 13]
```diff
@@ -47,14 +88,25 @@ async function mockSearchResponse( items ) {
...
+				{
+					type: 'page',
+					id: 1,
+					link: 'https://example.com/1',
+					title: {
+						rendered: 'My page',
+					},
+				},
+			],
+		} ),
```
Comment (tellthemachines) [inline packages/e2e-tests/specs/editor/blocks/navigation.test.js:109]: What's the purpose of these changes?

**Review by tellthemachines — COMMENTED** [review]
Thanks for persisting with this! It's almost there, just a few comments below.

File: packages/block-library/src/navigation-link/edit.js (hunk 12 above)
Comment (vcanales) [reply packages/block-library/src/navigation-link/edit.js]: Labelled by, label, etc, won't work directly on Tooltip text, as it's an out of flow dynamic element that never receives focus. I did find a hacky way of achieving this though:

https://github.com/WordPress/gutenberg/pull/31716/commits/a104888dbb10f1652691560dd921a301698b469f

A hidden element added to the span within `Tooltip`. This element has to be "hidden" with CSS in a particular way as to not remove it from the viewport, and instead making it "imperceptible" so that screen readers pick it up. We might want to find a way to achieve this from the Tooltip component itself though.

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js (hunk 13 above)
Comment (vcanales) [reply packages/e2e-tests/specs/editor/blocks/navigation.test.js:109]: When checking for the validity of the entity the navigation item links to, we have to call `getEntityRecord`, which throws a 404 error when it doesn't find an entity with the given parameters.

In this test suite, two tests are using the search box mocking only the search results, but not the underlying entities, which was causing `getEntityRecord` to throw the error, which would subsequently make the tests fail. There's a `createPages` function within this suite, but it didn't solve the problem, as mocked search results don't seem to care about entities that actually exist in the wordpress instance used for testing. Mocking the linked entities solved this problem.

File: packages/block-library/src/navigation-link/edit.js [hunk 14]
```diff
@@ -265,6 +266,64 @@ export const updateNavigationLinkBlockAttributes = (
...
+			missingText = __( 'Select category' );
+			break;
+		case 'tag':
+			/* translators: label for missing tag in navigation link block */
+			missingText = __( 'Select tag' );
+			break;
+		default:
+			/* translators: label for missing values in navigation link block */
+			missingText = __( 'Add link' );
+	}
```
Comment (spacedmonkey) [inline packages/block-library/src/navigation-link/edit.js:322]: These values shouldn't hardcoded. What about custom post type and taxonomies.

Comment (vcanales) [reply packages/block-library/src/navigation-link/edit.js:322]: Addressing this in a different PR, as it was only moved to a different place in this one.
